builder/build_validate.go

```diff
@@ -0,0 +1,911 @@
+package builder
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/clearlinux/mixer-tools/swupd"
+)
+
+type bundleStatus int
+
+const (
+	unchanged bundleStatus = iota
+	added
+	//removed
+	modified
+)
+
+// Contains all results from MCA diff and counters for added/deleted/modified bundles
+type mcaDiffResults struct {
+	bundleDiff map[string]*mcaBundleDiff
+
+	addCount int
+	delCount int
+	modCount int
+}
+
+// MCA diff results for a bundle
+type mcaBundleDiff struct {
+	name          string
+	status        bundleStatus
+	pkgFileCounts map[string]*bundlePkgStats
+
+	pkgFileDiffs diffLists
+	manFileDiffs diffLists
+	pkgDiffs     diffLists
+}
+
+// File counters for a package
+type bundlePkgStats struct {
+	add int
+	del int
+	mod int
+}
+
+// Lists of added, deleted, and modified packages or files
+type diffLists struct {
+	addList []string
+	delList []string
+	modList []string
+}
+
+// mcaBundleInfo contains manifest and package metadata necessary to perform an
+// MCA diff on a bundle.
+type mcaBundleInfo struct {
+	size uint64
+
+	subPkgs     map[string]bool
+	subPkgFiles map[string]*fileInfo
+	manFiles    map[string]*swupd.File
+}
+
+// mcaBundlePkgInfo contains MCA package and resolved file lists for a bundle
+// before subtraction.
+type mcaBundlePkgInfo struct {
+	allPkgs  map[string]*pkgInfo
+	allFiles map[string]bool
+}
+
+// pkgInfo contains package metadata
+type pkgInfo struct {
+	name    string
+	version string
+	arch    string
+
+	files map[string]*fileInfo
+}
+
+// fileInfo contains file metadata
+type fileInfo struct {
+	name  string
+	size  string
+	hash  string
+	modes string
+	links string
+	user  string
+	group string
+
+	pkg string
+}
+
+// CheckManifestCorrectness validates that the changes in manifest files between
+// two versions aligns to the corresponding RPM changes. Any mismatched files
+// between the manifests and RPMs will be printed as errors. When there are no
+// errors, package and file statistics for each modified bundle will be displayed.
+func (b *Builder) CheckManifestCorrectness(fromVer, toVer, downloadRetries int) error {
+	/* TODO
+	- Add bundle delete support
+	- Add format bump support
```

Comment:
Format bumps add more file exceptions where files are modified in manifests without a corresponding package change. Also, the current MCA does not support cross format comparisons, so I am planning on disabling MCA when the formats are different.

Reply:
You're right that the current MCA doesn't *support* cross-format comparisons, but they do occur in practice. During a format bump, MCA compares from +0 to +10 (same format), and from +0 to +20 (different formats). Then, the *reports* are compared to ensure that the expected changes appear:

- Because a new minversion is generally set during a format bump, the +0->+20 report indicates that a minversion bump occurred for every bundle in the distro.
- All other package/file adds, changes, and deletes should be identical for both +0->+10 and +0->+20 reports.
- Bundle deletes may occur, but are not reported about, as you have mentioned.

Reply:
Note that the report comparison I just described is not performed by MCA -- it's a separate check that's currently not automated. I wanted to mention it here though, because the report comparisons *have* caught format bump process issues in the past.

Reply:
Ok, I can update the checks for manifest files without package equivalents to account for formats. Also, I will need to account for /usr/share/defaults/swupd/format which is changed in the +10.

Reply:
It is challenging to compare special case files (files changed by manifest without package equivalent) across different formats because there is no way to determine whether a version is a +20 or an arbitrary version in the same format. The +20 requires different treatment than other versions, but it cannot be determined without tracking the first version in the format.

Based on offline discussion, MCA will correctly check special case files when comparing manifests within the same format and the +10 -> +20 comparison. When comparing manifests in separate formats (excluding the +10 -> +20 comparison), special case file errors will generate warning messages instead of errors.